YunoHost/yunohost
# Rework migration system to have independent migrations

## The problem

This follows discussions during meetings about https://github.com/YunoHost/yunohost/pull/585 which ships a critical migration (half of yunohost won't be able to ran as long as this migration ain't ran). 

The current migration system is too constraining as if a migration C or D fails (or has to be ran manually), then critical migration E won't be ran even though it's independent of C and D. 

So in the case of group permissions, the migration would / could not be ran just because some other pending / failed migration.

## Solution

The proposed solution is to rework the migration system to make migrations independent from each other (though we can still have dependencies between migration via `self.dependencies = [ ...list of required migrations to be done before this one ...]` inside the migration)

So if migrations C, D, and E are pending, running `yunohost tools migrations migrate` will attempt to run every of those pending migrations even if C or D fails.

This requires to change the way we store the state of migrations, as only the name/number of the last ran migrations was stored so far. Now instead, a yaml json containing the list of known migrations with a status `skipped`, `done` or `pending`.

In this PR, I also included many tweaks following past experiences with the migration mechanism, such as : 
- ability to re-run a migration even if it was already ran
- the `-t` option was confusing when you just want to revert the last migration (e.g. during development) so instead there's now a `--revert` option
- instead of json, yaml is probably better (i can remember many situation where I had to - or tell users to - manually edit the file for some reasons)

## PR Status

Tested on my side but should run more tests ...

Also this is blocking for 3.7 since it's a must have to be sure upgrades will happen smoothly even for people upgrading from old versions

## How to test

- Pull the branch
- If you run `yunohost tools migrations list` or `yunohost tools migrations state`, the old json should be transparently migrated to a new yaml file, you should be able to see this in `/etc/yunohost`
- Then you can try to revert some migrations such as `0007_ssh_conf_managed_by_yunohost_step1` and `0008_ssh_conf_managed_by_yunohost_step2` (though that last one you gotta do it by hand in the .yaml file because reverting aint supported for it)
- Then attempt to run just `0008_ssh_conf_managed_by_yunohost_step2`, should tell you that first you gotta run part1
- Then if you just run `yunohost tools migrations migrate` it should run both 0007 and 0008

... though there are many other options and use case to test, not sure I can list them all here é_è

## Validation
